storage/txn_cache.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2014 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: yananzhi (zac.zhiyanan@gmail.com)
+
+package storage
+
+import (
+	"sync"
+
+	"github.com/cockroachdb/cockroach/proto"
+	"github.com/cockroachdb/cockroach/util/cache"
+
+	gogoproto "github.com/gogo/protobuf/proto"
+)
+
+const (
+	// The default size of the txn cache.
+	defaultTxnCacheSize = 1 << 20
+)
+
+// TxnCache maintains txn which was pushed recently to avoid
```

Comment:
s/TxnCache/pushedTxnCache/. Rename the file to `pushed_txn_cache.go`.

I'd like to see some more explanatory comments about why this is necessary; I didn't really grasp the purpose of this cache until just now. Maybe something like this:
When two transactions conflict, they are likely to conflict on multiple keys. Each operation will separately see a conflicting intent, attempt to push the transaction, and resolve the intent. This cache can avoid an InternalPushTransaction RPC when we encounter a transaction that we have already attempted to push.